metal/resource_metal_port.go

```diff
@@ -0,0 +1,342 @@
+				Computed:    true,
+				Description: "Port type",
+			},
+			"mac": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "MAC address of the port",
+			},
+		},
+	}
+}
+
+func getPortByResourceData(d *schema.ResourceData, client *packngo.Client) (*packngo.Port, error) {
+	portId, portIdOk := d.GetOk("id")
+	deviceId, deviceIdOk := d.GetOk("device_id")
+	portName, portNameOk := d.GetOk("name")
+
+	var port *packngo.Port
+
+	if portIdOk && (deviceIdOk || portNameOk) {
+		return nil, fmt.Errorf("You must specify either id or (device_id and name)")
+	}
+	getOpts := &packngo.GetOptions{Includes: []string{
+		"native_virtual_network",
+		"virtual_networks",
+	}}
+	if portIdOk {
+		var err error
+		port, _, err = client.Ports.Get(portId.(string), getOpts)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		if !(deviceIdOk && portNameOk) {
+			return nil, fmt.Errorf("If you don't use port_id, you must supply both device_id and name")
+		}
+		device, _, err := client.Devices.Get(deviceId.(string), getOpts)
+		if err != nil {
+			return nil, err
+		}
+		port, err = device.GetPortByName(portName.(string))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return port, nil
+}
+
+func getSpecifiedNative(d *schema.ResourceData) string {
+	nativeRaw, nativeOk := d.GetOk("native_vlan_id")
+	specifiedNative := ""
+	if nativeOk {
+		specifiedNative = nativeRaw.(string)
+	}
+	return specifiedNative
+}
+
+func getCurrentNative(p *packngo.Port) string {
+	currentNative := ""
+	if p.NativeVirtualNetwork != nil {
+		currentNative = p.NativeVirtualNetwork.ID
+	}
+	return currentNative
+}
+
+func attachedVlanIds(p *packngo.Port) []string {
+	attached := []string{}
+	for _, v := range p.AttachedVirtualNetworks {
+		attached = append(attached, v.ID)
+	}
+	return attached
+}
+
+func specifiedVlanIds(d *schema.ResourceData) []string {
+	vlanIdsRaw, vlanIdsOk := d.GetOk("vlan_ids")
+	specified := []string{}
+	if vlanIdsOk {
+		specified = convertStringArr(vlanIdsRaw.([]interface{}))
+	}
+	return specified
+}
+
+type portVlanAction func(*packngo.PortAssignRequest) (*packngo.Port, *packngo.Response, error)
+
+func processVlansOnPort(portId string, vlanIds []string, f portVlanAction) error {
+	par := packngo.PortAssignRequest{PortID: portId}
+	for _, vId := range vlanIds {
+		par.VirtualNetworkID = vId
+		_, _, err := f(&par)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+/*
+Create and Update will be probably identical in case of metal_port
+
+I figured we should process port action in order:
+ - native VLAN detachment
+ - VLANs detachment
+ - bond break
+ - l2/l3 conversion
+ - bond creation
+ - VLANs assignment
+ - native VLAN assignment
+*/
+
+func resourceMetalPortCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*packngo.Client)
+
+	port, err := getPortByResourceData(d, client)
+	if err != nil {
+		return err
+	}
+
+	// Constraint: Only bond ports have layer2 mode
```

Comment:
Each of these commented sections could be its own function to keep the cyclomatic complexity down.  Your comment block at https://github.com/equinix/terraform-provider-metal/pull/116/files#diff-266370b6fe0d745e5c92c5193546cb0a9f8d0c56ef0b65355e63759fc44b2b92R188-R195 would become a function calling each of these functions.